File: 332-reconstruct-itinerary/reconstruct-itinerary.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        mapper = {a: [] for a, b in tickets}
        for i, j in tickets:
            mapper[i].append(j)
        
        for k in mapper:
            mapper[k].sort()

        
        result = []
        
        def dfs(m, node):
            if node in m:
                temp = m[node][:]
                while temp:
                    t = temp[0]
                    m[node].pop(0)
                    dfs(m, t)
                    temp = m[node][:]
            result.append(node)

        dfs(mapper, "JFK")
        
        result.reverse()

        return result if len(result) == (len(tickets) + 1) else []

File: 332-reconstruct-itinerary/reconstruct-itinerary.java
class Solution {
    public void dfs(HashMap<String, List<String>> m, List<String> result, String node){
        if (m.containsKey(node)){
            List<String> temp = m.get(node);
            while (temp.size()>0){
                String t = temp.get(0);
                temp.remove(0);
                m.put(node, temp);
                dfs(m, result, t);
            }
        }
        result.add(node);
    }

    public List<String> findItinerary(List<List<String>> tickets) {
        HashMap<String, List<String>> mapper = new HashMap<String, List<String>>();
        for (List<String> temp: tickets){
            List<String> t = mapper.getOrDefault(temp.get(0), new ArrayList<String>());
            t.add(temp.get(1));
            mapper.put(temp.get(0), t);
        }

        for (Map.Entry<String, List<String>>i: mapper.entrySet()){
            Collections.sort(mapper.get(i.getKey()));
        }

        List<String> result = new ArrayList<String>();
        dfs(mapper, result, "JFK");

        Collections.reverse(result);
        return result.size() == tickets.size() + 1 ? result : new ArrayList<String>();         
    }
}